### ISSUE TITLE
storage: protect against raft proposal replay/reordering

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#6287]
See #6166.

Implement protection against Raft replays/reordering, ideally such that the leader knows at proposal time the log index at which commands commit (unless they have to be reproposed).

From #6166:

Replay protection is naively an issue because applying a logicless batch of writes is at odds with our current method, which largely relies on write/timestamp conflicts. However, as @bdarnell pointed out, as of recently (#5973) we (mostly) colocate the Raft leader and the leader lease holder on the same node and have more options for dealing with this (i.e. making the application of a command conditional on the Raft log position it was originally proposed at).

More concretely (via @bdarnell):

The RaftCmd would also include a term and log index, to be populated with the leader's latest log information when it is proposed. The writes would only be applied if it committed at that index; they would be dropped (and reproposed) if they committed at any other index. This would only be a possibility when the lease holder is not the raft leader (and even then it's unlikely in stable lease situations). We may be able to be even stricter about colocating the two roles (in which case we could perhaps remove raft-level replays completely), but I'm wary of forcing too many raft elections.

**Comment by tschottdorf** [comment]
This issue should also include auditing/removing all of the current replay-protection code which may then be obsolete.

**Comment by tschottdorf** [comment]
Implemented in #6961. Regarding the audit, we may not need additional replay protection for Raft any more, but we've used the MVCC Write-too-old protection to fade out most of the abort cache usage, so that replays from the `DistSender` (i.e. resending a message after a timeout) look just like a Raft replay (the only difference is the higher sequence number, and that isn't checked outside of the abort cache). Hence, nothing to remove as far as I can tell.
